elf_reader.go

```diff
@@ -279,86 +282,139 @@ func (ec *elfCode) loadPrograms() (map[string]*ProgramSpec, error) {
+			delete(progs, n)
 		}
 	}
 
-	res := make(map[string]*ProgramSpec, len(progs))
-	for _, prog := range progs {
-		err := link(prog, libs)
+	return progs, nil
+}
+
+// loadFunctions extracts instruction streams from the given program section
+// starting at each symbol in the section. The section's symbols must already
+// be narrowed down to STT_NOTYPE (emitted by clang <8) or STT_FUNC.
+//
+// The resulting map is indexed by function name.
+func (ec *elfCode) loadFunctions(section *elfSection) (map[string]asm.Instructions, error) {
+	funcs := make(map[string]asm.Instructions)
+	for _, sym := range section.symbols {
+		// Load the function's instructions. Pass in the absolute offset of the
+		// function within the section to allow resolving relocations.
+		si, err := ec.loadInstructions(section, sym)
 		if err != nil {
-			return nil, fmt.Errorf("program %s: %w", prog.Name, err)
+			return nil, fmt.Errorf("loading instructions: %w", err)
 		}
-		res[prog.Name] = prog
+
+		funcs[sym.Name] = si
 	}
 
-	return res, nil
+	return funcs, nil
 }
 
-func (ec *elfCode) loadInstructions(section *elfSection) (asm.Instructions, uint64, error) {
-	var (
-		r      = bufio.NewReader(section.Open())
-		insns  asm.Instructions
-		offset uint64
-	)
+// loadInstructions extracts the instruction stream of the given symbol
+// from the given ELF section.
+func (ec *elfCode) loadInstructions(section *elfSection, sym elf.Symbol) (asm.Instructions, error) {
+	// clang up until at least version 7 does not set the symbol's size field.
+	// In this case, buffer the remainder of the section, since we only read up
+	// to the first Exit instruction.
+	if sym.Size == 0 {
+		// Don't use Max(U)Int64, the offset is added to this number
+		// during SectionReader creation and will overflow.
+		sym.Size = section.Size
+	}
+
+	r := internal.NewBufferedSectionReader(section, int64(sym.Value), int64(sym.Size))
+
+	// Starting point of the function within the section.
+	offset := sym.Value
+
+	var insns asm.Instructions
 	for {
 		var ins asm.Instruction
 		n, err := ins.Unmarshal(r, ec.ByteOrder)
-		if err == io.EOF {
-			return insns, offset, nil
-		}
 		if err != nil {
-			return nil, 0, fmt.Errorf("offset %d: %w", offset, err)
+			// EOF is an error, a valid instruction stream ends with an Exit instruction.
+			return nil, fmt.Errorf("offset %d: %w", offset, err)
 		}
 
 		ins.Symbol = section.symbols[offset].Name
 
+		// Up to clang9, jumps to subprograms within the same ELF section were
+		// encoded using offsets relative to the instruction. When parsing
+		// subprograms individually, other subprograms within the same section
+		// are no longer in scope, so these jump offsets are no longer meaningful.
+		// Look up the destination in the section's symbol table, store its
+		// reference by name and blind the relative jump offset. The linker will
+		// figure out what this jump points to later.
+		if ins.IsFunctionCall() && ins.Constant != -1 {
```

Comment:
True, though we don't test fp loads using pre-clang9 at the moment. Is this something we should consider? clang9 is quite recent. I've pushed a commit that merges `IsFunctionCall()` and `IsLoadOfFunctionPointer()` since telling the difference is currently not necessary: 0406342ef132082bf368b86ede01ebcb69ca34e1. If we need this at some point, we can revert, but for now it makes things somewhat harder to understand.

Reply:
Please keep them as they are, `asm` is public and these are two quite different things.

Reply:
Done, added `IsFunctionReference()` in fab3ad25ce2f01b039f52dee98afe8070f2d502d.